### PR TITLE
Verify that disk volumes are mounted by UUID in fstab

### PR DESCRIPTION
**Pull request by pcahyna**
Note that disk volumes mean filesystems directly on disks.

Currently the test fails: on first invocation the filesystem gets mounted by device name and only on the second by UUID. This should be also caught by a real idempotence test.

XXX this should be part of the existing verification task lists: there is already a task called "Verify that the device identifier appears in /etc/fstab", but it does not check for this condition.

### CODE REVIEW

**Comment by pcahyna** [comment]
here's what happens:
One invocation has created a fstab entry which mounts the filesystem by block device path:
```
TASK [storage : manage mounts to match the specified state] ********************
changed: [/home/pcahyna/linux-system-roles-testing/local/rhel-guest-image-7.7-200.x86_64.qcow2] => (item={'path': '/opt/test', 'state': 'absent'}) => {"ansible_loop_var": "mount_info", "changed": true, "dump": "0", "fstab": "/etc/fstab", "mount_info": {"path": "/opt/test", "state": "absent"}, "name": "/opt/test", "opts": "defaults", "passno": "0"}
changed: [/home/pcahyna/linux-system-roles-testing/local/rhel-guest-image-7.7-200.x86_64.qcow2] => (item={'src': '/dev/vdc', 'dump': 0, 'passno': 0, 'fstype': 'ext4', 'state': 'mounted', 'path': '/opt/test', 'opts': 'defaults'}) => {"ansible_loop_var": "mount_info", "changed": true, "dump": "0", "fstab": "/etc/fstab", "fstype": "ext4", "mount_info": {"dump": 0, "fstype": "ext4", "opts": "defaults", "passno": 0, "path": "/opt/test", "src": "/dev/vdc", "state": "mounted"}, "name": "/opt/test", "opts": "defaults", "passno": "0", "src": "/dev/vdc"}
```
...
```
TASK [Read the /etc/fstab file for volume existence] ***************************
ok: [/home/pcahyna/linux-system-roles-testing/local/rhel-guest-image-7.7-200.x86_64.qcow2] => {"changed": false, "cmd": ["cat", "/etc/fstab"], "delta": "0:00:00.003845", "end": "2019-10-15 13:43:52.449118", "rc": 0, "start": "2019-10-15 13:43:52.445273", "stderr": "", "stderr_lines": [], "stdout_lines": ["/dev/vdc /opt/test ext4 defaults 0 0"]}
```
and the repeated invocation changes it to mount by UUID:
```
TASK [storage : manage mounts to match the specified state] ********************
changed: [/home/pcahyna/linux-system-roles-testing/local/rhel-guest-image-7.7-200.x86_64.qcow2] => (item={'src': 'UUID=26bddec8-c250-416c-b8f1-7d8bfc070d06', 'dump': 0, 'passno': 0, 'fstype': 'ext4', 'state': 'mounted', 'path': '/opt/test', 'opts': 'defaults'}) => {"ansible_loop_var": "mount_info", "changed": true, "dump": "0", "fstab": "/etc/fstab", "fstype": "ext4", "mount_info": {"dump": 0, "fstype": "ext4", "opts": "defaults", "passno": 0, "path": "/opt/test", "src": "UUID=26bddec8-c250-416c-b8f1-7d8bfc070d06", "state": "mounted"}, "name": "/opt/test", "opts": "defaults", "passno": "0", "src": "UUID=26bddec8-c250-416c-b8f1-7d8bfc070d06"}
```
...
```
TASK [Read the /etc/fstab file for volume existence] ***************************
ok: [/home/pcahyna/linux-system-roles-testing/local/rhel-guest-image-7.7-200.x86_64.qcow2] => {"changed": false, "cmd": ["cat", "/etc/fstab"], "delta": "0:00:00.004238", "end": "2019-10-15 13:45:30.693829", "rc": 0, "start": "2019-10-15 13:45:30.689591", "stderr": "", "stderr_lines": [], "stdout_lines": ["UUID=26bddec8-c250-416c-b8f1-7d8bfc070d06 /opt/test ext4 defaults 0 0"]}
```
(copied from https://github.com/linux-system-roles/storage/commit/36da735ad34a9b7c0cd2067adbd8124eef1641d4#r35511810 )

**Comment by dwlehman** [comment]
The rules are a bit more complicated. Device-mapper devices are generally listed by path instead of by UUID since that path can only change as a direct result of the user changing it, as opposed to partitions or disks which can change arbitrarily depending on device detection ordering.

I am definitely open to verifying that the correct identifier is used, but it will have to be a bit more complex than this.

File: tests/tests_change_disk_fs.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@
     mount_location: '/opt/test'
     volume_size: '5g'
     fs_type_after: "{{ 'ext3' if (ansible_distribution == 'RedHat' and ansible_distribution_major_version == '6') else 'ext4' }}"
-
+    pat: "{{ '^([^#\\s]+)\\s+' + mount_location + '\\s.*' }}"
   tasks:
     - include_role:
         name: storage
```
Comment (richm) [inline tests/tests_change_disk_fs.yml:12]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+- hosts: all
+  become: true
+  vars:
+    storage_safe_mode: false
+    mount_location: '/opt/test1'
+    volume_size: '5g'
+
+  tasks:
+    - include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:11]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        max_return: 1
+
+    ##
+    ## Disk
+    ##
+    - name: Test for correct handling of new encrypted volume w/ no key
+      block:
+        - name: Create an encrypted disk volume w/ default fs
+          include_role:
+            name: storage
```
Comment (richm) [inline tests/tests_luks.yml:25]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+          assert:
+            that: "blivet_output.failed and
+                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
+                   not blivet_output.changed"
+            msg: "Unexpected behavior w/ encrypted pool w/ no key"
+
+    # encrypted disk volume
+    - name: Create an encrypted disk volume w/ default fs
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:55]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            disks: "{{ unused_disks }}"
+            mount_point: "{{ mount_location }}"
+            encryption: true
+            encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Remove the encryption layer
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:69]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+            disks: "{{ unused_disks }}"
+            mount_point: "{{ mount_location }}"
+            encryption: false
+            encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Add encryption to the volume
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:83]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+    ##
+    ## Partition
+    ##
+
+    - name: Test for correct handling of new encrypted volume w/ no key
+      block:
+        - name: Create an encrypted partition volume w/ default fs
+          include_role:
+            name: storage
```
Comment (richm) [inline tests/tests_luks.yml:103]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        - name: Verify the output of the keyless luks test
+          assert:
+            that: "blivet_output.failed and
+                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
+                   not blivet_output.changed"
+            msg: "Unexpected behavior w/ encrypted pool w/ no key"
+
+    - name: Create an encrypted partition volume w/ default fs
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:136]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+                mount_point: "{{ mount_location }}"
+                size: 4g
+                encryption: true
+                encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Remove the encryption layer
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:154]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+                mount_point: "{{ mount_location }}"
+                size: 4g
+                encryption: false
+                encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Add encryption to the volume
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:172]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+    ##
+    ## LVM
+    ##
+
+    - name: Test for correct handling of new encrypted volume w/ no key
+      block:
+        - name: Create an encrypted lvm volume w/ default fs
+          include_role:
+            name: storage
```
Comment (richm) [inline tests/tests_luks.yml:196]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+        - name: Verify the output of the keyless luks test
+          assert:
+            that: "blivet_output.failed and
+                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
+                   not blivet_output.changed"
+            msg: "Unexpected behavior w/ encrypted pool w/ no key"
+
+    - name: Create an encrypted lvm volume w/ default fs
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:228]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+                mount_point: "{{ mount_location }}"
+                size: 4g
+                encryption: true
+                encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Remove the encryption layer
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:245]: please use `name: linux-system-roles.storage`

File: tests/tests_luks.yml [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+                mount_point: "{{ mount_location }}"
+                size: 4g
+                encryption: false
+                encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Add encryption to the volume
+      include_role:
+        name: storage
```
Comment (richm) [inline tests/tests_luks.yml:262]: please use `name: linux-system-roles.storage`

**Comment by richm** [comment]
should we close this in favor of #104 ?

**Comment by richm** [comment]
[citest pending]

**Comment by richm** [comment]
[citest bad]

**Comment by richm** [comment]
[citest pending]

File: tests/tests_luks.yml [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+                mount_point: "{{ mount_location }}"
+                size: 4g
+                encryption: true
+                encryption_passphrase: 'yabbadabbadoo'
+
+    - include_tasks: verify-role-results.yml
+
+    - name: Clean up
+      include_role:
+        name: storage
```
Comment (nhosoi) [inline tests/tests_luks.yml:279]: One more?

> please use name: linux-system-roles.storage

**Comment by richm** [comment]
closing due to inactivity

**Pull request closed without merging**